### PR TITLE
HTTP implementation of voluntary exit pool endpoints

### PR DESCRIPTION
**Pull request by rkapka**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

HTTP handlers for https://ethereum.github.io/beacon-APIs/?urls.primaryName=dev#/Beacon/getPoolVoluntaryExits and https://ethereum.github.io/beacon-APIs/?urls.primaryName=dev#/Beacon/submitPoolVoluntaryExit

Benchmarks shows 1ms for both endpoints, although currently there are no exits in the pool. 

**Which issues(s) does this PR fix?**

Part of #12635

### CODE REVIEW

File: beacon-chain/rpc/eth/beacon/handlers_pool.go [hunk 1]
```diff
@@ -168,3 +171,85 @@ func (s *Server) SubmitAttestations(w http.ResponseWriter, r *http.Request) {
...
+	http2.WriteJson(w, &ListVoluntaryExitsResponse{Data: exits})
+}
+
+// SubmitVoluntaryExit submits SignedVoluntaryExit object to node's pool
+// and if passes validation node MUST broadcast it to network.
+func (s *Server) SubmitVoluntaryExit(w http.ResponseWriter, r *http.Request) {
+	ctx, span := trace.StartSpan(r.Context(), "beacon.SubmitVoluntaryExit")
+	defer span.End()
+
+	if r.Body == http.NoBody {
```
Comment (james-prysm) [inline beacon-chain/rpc/eth/beacon/handlers_pool.go:200]: good chance to use that new technique 
```
	var jsonRegistrations []*shared.SignedValidatorRegistration
	err := json.NewDecoder(r.Body).Decode(&jsonRegistrations)
	switch {
	case err == io.EOF:
		http2.HandleError(w, "No data submitted", http.StatusBadRequest)
		return
	case err != nil:
		http2.HandleError(w, "Could not decode request body: "+err.Error(), http.StatusBadRequest)
		return
	}
	```
 as an example

File: beacon-chain/rpc/eth/beacon/handlers_pool.go [hunk 2]
```diff
@@ -168,3 +171,85 @@ func (s *Server) SubmitAttestations(w http.ResponseWriter, r *http.Request) {
...
+	}
+	validate := validator.New()
+	if err := validate.Struct(req); err != nil {
+		http2.HandleError(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	exit, err := req.ToConsensus()
+	if err != nil {
+		http2.HandleError(w, "Could not convert request exit to consensus exit: "+err.Error(), http.StatusBadRequest)
```
Comment (saolyn) [inline beacon-chain/rpc/eth/beacon/handlers_pool.go:217]:
```suggestion
		http2.HandleError(w, "Could not convert exit request to consensus exit: "+err.Error(), http.StatusBadRequest)
```

File: beacon-chain/rpc/eth/shared/structs.go [hunk 3]
```diff
@@ -82,10 +82,20 @@ type ValidatorRegistration struct {
 }
 
 type SignedValidatorRegistration struct {
-	Message   *ValidatorRegistration `json:"message" validate:"required,dive"`
+	Message   *ValidatorRegistration `json:"message" validate:"required"`
```
Comment (james-prysm) [inline beacon-chain/rpc/eth/shared/structs.go:85]: nice catch here not sure why that's set there...

File: beacon-chain/rpc/eth/shared/structs.go [hunk 4]
```diff
@@ -383,6 +393,52 @@ func (b *BeaconCommitteeSubscription) ToConsensus() (*validator.BeaconCommitteeS
...
+		return nil, NewDecodeError(err, "Signature")
+	}
+	exit, err := e.Message.ToConsensus()
+	if err != nil {
+		return nil, NewDecodeError(err, "Message")
+	}
+
+	return &eth.SignedVoluntaryExit{
+		Exit:      exit,
+		Signature: sig,
```
Comment (james-prysm) [inline beacon-chain/rpc/eth/shared/structs.go:408]: should we check length here

File: beacon-chain/rpc/eth/beacon/handlers_pool.go (hunk 2 above)
Comment (rkapka) [reply beacon-chain/rpc/eth/beacon/handlers_pool.go:217]: Elsewhere I use the same sentence structure, so I would prefer keeping it like this.

File: beacon-chain/rpc/eth/beacon/handlers_pool.go (hunk 1 above)
Comment (rkapka) [reply beacon-chain/rpc/eth/beacon/handlers_pool.go:200]: I will rewrite this everywhere

**Review by james-prysm — APPROVED** [review]

**Review by james-prysm — APPROVED** [review]

**Pull request merged**